File: strategies/momentum.py
import pandas as pd
import numpy as np
from .base import Strategy


class MomentumStrategy(Strategy):
    def __init__(self, lookback=20):
        super().__init__('Momentum')
        self.lookback = lookback

    def generate_signals(self, data):
        df = data.copy()

        # calculate momentum as percent change over lookback period
        df['momentum'] = df['close'].pct_change(periods=self.lookback)

        df['signal'] = 0
        df.loc[df['momentum'] > 0, 'signal'] = 1
        df.loc[df['momentum'] < 0, 'signal'] = -1

        # TODO: add threshold so we dont trade on tiny momentum
        # TODO: maybe add volume confirmation

        return df

    def get_params(self):
        return {'lookback': self.lookback}


# TODO: implement mean reversion strategy
# TODO: bollinger bands strategy

File: lib/strategies/momentum.ts
/**
 * Momentum Strategy
 * Trend-following based on recent price momentum
 */

import { Strategy, OHLCV, SignalData, StrategyParams } from './types';
import { pctChange } from '../math/rolling';

export interface MomentumParams {
  lookback: number;
}

export class MomentumStrategy implements Strategy {
  name = 'Momentum';
  private lookback: number;

  constructor(params: Partial<MomentumParams> = {}) {
    this.lookback = params.lookback ?? 20;
  }

  generateSignals(data: OHLCV[]): SignalData[] {
    const closes = data.map((d) => d.close);

    // Calculate momentum as percent change over lookback period
    const momentum = pctChange(closes, this.lookback);

    const signals: SignalData[] = data.map((row, i) => {
      let signal: -1 | 0 | 1 = 0;
      const m = momentum[i];

      if (!Number.isNaN(m)) {
        if (m > 0) signal = 1; // Positive momentum - buy
        else if (m < 0) signal = -1; // Negative momentum - sell
      }

      return {
        ...row,
        signal,
        momentum: m,
      };
    });

    return signals;
  }

  getParams(): StrategyParams {
    return {
      lookback: this.lookback,
    };
  }
}
